### PR TITLE
[GR-50220] Allow resources inclusion from jar files in paths containing spaces

### PR DESCRIPTION
**Pull request by zakkak**
Since https://github.com/oracle/graal/pull/7670 it is no longer possible to build projects that include resources from paths with spaces.

How to reproduce:

1. Build graal `master` from source
2. `export GRAALVM_HOME=/path/to/graal/master/build`
3. `git clone https://github.com/quarkusio/quarkus`
4. `cd quarkus`
5. `./mvnw -Dquickly`
6. `./mvnw -Dnative -pl integration-tests/simple\ with\ space -Dnative.surefire.skip -Dformat.skip -Dno-descriptor-tests clean verify -Dquarkus.native.container-build=false`

This PR resolves this issue and adds a set of tests for resource names with spaces (which don't catch the issue at hand, but are still good to test against).

Closes https://github.com/quarkusio/quarkus/issues/36998

### CODE REVIEW

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ResourcesFeature.java [hunk 1]
```diff
@@ -343,17 +343,17 @@ private void registerResource(Module module, String resourcePath, boolean fromJa
...
+                return ((JarURLConnection) url.openConnection()).getJarFile();
             } catch (IOException e) {
                 throw new RuntimeException(e);
             }
         }
 
         private boolean resourceIsDirectory(URL url, boolean fromJar, String resourcePath) throws IOException, URISyntaxException {
             if (fromJar) {
-                try (JarFile jf = new JarFile(urlToJarPath(url))) {
+                try (JarFile jf = urlToJarFile(url)) {
```
[the remark this answers is not in the document]
Comment (zakkak) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ResourcesFeature.java]: > I see that it [creates a copy of the file](https://github.com/openjdk/jdk/blob/2e34a2ebf0f14043b129461b0397495e7e75a38b/src/java.base/share/classes/sun/net/www/protocol/jar/URLJarFile.java#L216), which is probably not what we want.

Good catch @jerboaa 

> So I suggest to use...

Done. Thanks!

**Review by jerboaa — APPROVED** [review]
Would be nice to actually catch this in a test, but this seems fine to me and should be a safe change. Thanks!

**Comment by zakkak** [comment]
> Would be nice to actually catch this in a test

One option would be to build graal in a folder containing a space, but unfortunately that's not supported (there are some ninja related failures if you try to do it). 

Another option would be to add a space in the jar file itself, the name is derived from https://github.com/oracle/graal/blob/35d412e56ab312c7bebd093c0ae1f5ef3092e6da/substratevm/mx.substratevm/suite.py#L1978 so the following patch does the trick:

```diff
diff --git a/substratevm/mx.substratevm/suite.py b/substratevm/mx.substratevm/suite.py
index 7c9555bccb5..713abcb06f4 100644
--- a/substratevm/mx.substratevm/suite.py
+++ b/substratevm/mx.substratevm/suite.py
@@ -1975,7 +1975,7 @@ suite = {
             "testDistribution" : True,
         },
 
-        "SVM_TESTS" : {
+        "SVM TESTS" : {
           "subDir": "src",
           "relpath" : True,
           "dependencies" : [
```

However, it will apply to all tests, i.e. we won't have a separate jar with spaces in the filename only for this test case. Which means that without this PR it results in:

```
com.oracle.svm.configure.test.config.OmitPreviousConfigTests started (4 of 65)
  testConfigDifference: Fatal error: Unexpected error while locating the configuration files.: java.lang.RuntimeException: java.nio.file.NoSuchFileException: /home/zakkak/code/graal/substratevm/mxbuild/jdk22/dists/jdk21/svm%20tests.jar
    at com.oracle.svm.core.jdk.Resources.get(Resources.java:317)
    at com.oracle.svm.core.jdk.Resources.createURLs(Resources.java:434)
    at com.oracle.svm.core.jdk.Resources.createURL(Resources.java:366)
    at com.oracle.svm.core.jdk.Resources.createURL(Resources.java:358)
    at com.oracle.svm.core.jdk.ResourcesHelper.nameToResourceURL(ResourcesHelper.java:102)
    at jdk.internal.loader.BootLoader.findResource(BootLoader.java:609)
    at java.lang.ClassLoader.getResource(ClassLoader.java:1409)
    at java.lang.ClassLoader.getResource(ClassLoader.java:1407)
    at java.lang.Class.getResource(DynamicHub.java:3160)
    at com.oracle.svm.configure.test.config.OmitPreviousConfigTests.lambda$loadTraceProcessorFromResourceDirectory$0(OmitPreviousConfigTests.java:68)
    at com.oracle.svm.configure.config.ConfigurationFileCollection.addDirectory(ConfigurationFileCollection.java:73)
    at com.oracle.svm.configure.test.config.OmitPreviousConfigTests.loadTraceProcessorFromResourceDirectory(OmitPreviousConfigTests.java:65)
    at com.oracle.svm.configure.test.config.OmitPreviousConfigTests.testConfigDifference(OmitPreviousConfigTests.java:107)
    at java.lang.reflect.Method.invoke(Method.java:580)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.junit.runners.Suite.runChild(Suite.java:128)
    at org.junit.runners.Suite.runChild(Suite.java:27)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
    at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
    at com.oracle.mxtool.junit.MxJUnitWrapper.runRequest(MxJUnitWrapper.java:376)
    at com.oracle.svm.junit.SVMJUnitRunner.run(SVMJUnitRunner.java:164)
    at com.oracle.svm.junit.SVMJUnitRunner.main(SVMJUnitRunner.java:184)
```

I will try to create a new jar file (mx distribution) and add it as a dependency to `SVM_JAR` to make this test-specific.

**Comment by zakkak** [comment]
> I will try to create a new jar file (mx distribution) and add it as a dependency to SVM_TESTS to make this test-specific.

Done in https://github.com/oracle/graal/pull/7815/commits/a24ec9aa3928980f37418e51a3f24f0f9e968b72

File: substratevm/mx.substratevm/suite.py [hunk 2]
```diff
@@ -1992,6 +2010,19 @@
           "testDistribution" : True,
         },
 
+        "SVM_TESTS WITH SPACE" : {
```
Comment (jerboaa) [inline substratevm/mx.substratevm/suite.py:2013]: Might warrant a comment that there is **intentionally** a space in the name. Hard to spot ;-)

**Review by jerboaa — APPROVED** [review]
LGTM. @fniephaus Could you help us get this pushed/reviewed? Thanks!

**Comment by jerboaa** [comment]
@olpaw @dnestoro Could you help us getting this reviewed/integrated, please? It's a regression caused by https://github.com/oracle/graal/pull/7670. Thanks!

**Comment by jerboaa** [comment]
> > @olpaw @dnestoro Could you help us getting this reviewed/integrated, please? It's a regression caused by #7670. Thanks!
> 
> Sure, I will take a look on the PR next week (today I have some other priorities). Is that okay (or it is urgent to be done today)?

Next week is fine. We just don't want such issues hanging in the air without knowing if you guys are aware (as it's causing test failures for us). Thanks for the reply!

**Review by dnestoro — APPROVED** [review]

**Comment by dnestoro** [comment]
Looks good to me! @olpaw can you also take a look on it before it gets merged?

File: substratevm/mx.substratevm/suite.py (hunk 2 above)
Comment (olpaw) [reply substratevm/mx.substratevm/suite.py:2013]: Yes indeed. @zakkak please add a comment in `suite.py` why we need exactly that.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ResourcesFeature.java [hunk 3]
```diff
@@ -345,8 +345,8 @@ private void registerResource(Module module, String resourcePath, boolean fromJa
         /* Util functions for resource attributes calculations */
         private String urlToJarPath(URL url) {
             try {
-                return ((JarURLConnection) url.openConnection()).getJarFileURL().getFile();
-            } catch (IOException e) {
+                return ((JarURLConnection) url.openConnection()).getJarFileURL().toURI().getPath();
+            } catch (IOException | URISyntaxException e) {
```
Comment (olpaw) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ResourcesFeature.java:349]: Thanks for the small cleanups as well :grin:

**Review by olpaw — APPROVED** [review]

File: substratevm/mx.substratevm/suite.py (hunk 2 above)
Comment (zakkak) [reply substratevm/mx.substratevm/suite.py:2015]: Done.

**Comment by zakkak** [comment]
@jerboaa @dnestoro @olpaw Thank you for the review

**Comment by jerboaa** [comment]
Am I correct in thinking that this is in the merge queue?

**Comment by olpaw** [comment]
Running in internal CI. Should be in merge queue soon.

**Comment by jerboaa** [comment]
@olpaw Any update on the merge queue status of this? Did it fail?

**Comment by zapster** [comment]
@jerboaa it is still in the merge queue. There is a lot going on currently, but I'm having an eye on it and make sure that it will go in.

**Comment by jerboaa** [comment]
@zapster Thank you! Appreciate that.

**Pull request merged**